src/generators/javascript/templates/index.hbs

```diff
@@ -327,12 +423,363 @@ export function {{functionName}}(
 {{/if}}
 {{/each}}
 
+{{#each page}}
+{{> functionDocumentation isBrowser=../isBrowser event="page"}}
+{{#if ../isBrowser}}
+export function {{functionName}}(
+	category: string,
+	name: string,
+	props{{#if isPropertiesOptional}}?{{/if}}: {{propertiesType}},
+	options?: ApiOptions,
+	callback?: apiCallback,
+): void;
+export function {{functionName}}(
+  name: string,
+  props{{#if isPropertiesOptional}}?{{/if}}: {{propertiesType}},
+  options?: ApiOptions,
+  callback?: apiCallback,
+): void;
+export function {{functionName}}(
+	props{{#if isPropertiesOptional}}?{{/if}}: {{propertiesType}},
+	options?: ApiOptions,
+	callback?: apiCallback,
+): void;
+export function {{functionName}}(
+  categoryOrProps?: string | {{propertiesType}},
+  nameOrPropsOrOptions?: string | {{propertiesType}} | ApiOptions,
+  propsOrOptionsOrCallback?: {{propertiesType}} | ApiOptions | apiCallback,
+  optionsOrCallback?: ApiOptions | apiCallback,
+  callback?: apiCallback,
+): void {
+	const a = analytics();
+	{{#if ../isDevelopment}}
+	const schema = {{rawJSONSchema}}
+	const message = {
+		event: '{{rawEventName}}',
+		properties: props || {},
+		options,
+	}
+	validateAgainstSchema(message, schema).then(()=>{
+		if (a) {
+			if (typeof categoryOrProps === 'string' && typeof nameOrPropsOrOptions === 'string') {
+				// Overload (category: string, name: string, props?: {{propertiesType}}, options?: ApiOptions, callback?: apiCallback)
+				a.page(
+					categoryOrProps,
+					'{{rawEventName}}',
+					(propsOrOptionsOrCallback as {{propertiesType}}) || {},
+					withRudderTyperContext(optionsOrCallback as ApiOptions),
+					callback,
+				);
+			} else if (typeof categoryOrProps === 'string') {
+				// Overload (name: string, props?: {{propertiesType}}, options?: ApiOptions, callback?: apiCallback)
+				a.page(
+					'{{rawEventName}}',
+					(nameOrPropsOrOptions as {{propertiesType}}) || {},
+					withRudderTyperContext(propsOrOptionsOrCallback as ApiOptions),
+					optionsOrCallback as apiCallback,
+				);
+			} else {
+				// Overload (props?: {{propertiesType}}, options?: ApiOptions, callback?: apiCallback)
+				a.page(
+					categoryOrProps || {},
+					withRudderTyperContext(nameOrPropsOrOptions as ApiOptions),
+					propsOrOptionsOrCallback as apiCallback,
+				);
+			}
+		}
+	})
+	{{else}}
+	if (a) {
```

Comment:
Instead of repeating this block, we should extract into an inline function that can be reused.

Reply:
Let's do the same for other APIs as well.

Reply:
In handlebars language we cannot define inline function I guess

Reply:
It's ultimately TS code within the handlebar template. I'm sure you can do it. Let's give it a try.

Reply:
Done